runtime/src/registrar.rs

```diff
@@ -0,0 +1,355 @@
+			}
+		}
+
+		/// Block initializer. Clears SelectedThreads and constructs/replaces Active.
+		fn on_initialize() {
+			let paras = Parachains::get().into_iter()
+				.map(|id| (id, None))
+				.chain(SelectedThreads::take().into_iter()
+					.map(|(who, id)| (who, Some(id))
+				))
+				.map(|(id, x)| (id, x, ParaInfo::get(x).origins))
+				.collect::<Vec<_>>();
+			Active::put(paras);
+		}
+	}
+}
+
+pub enum Events<T: Trait> {
+	/// A parathread was registered; its new ID is supplied.
+	ParathreadRegistered(ParaId),
+
+	/// The parathread of the supplied ID was de-registered.
+	ParathreadDeregistered(ParaId),
+}
+
+impl<T: Trait> Module<T> {
+	pub fn ensure_thread_id(id: ParaId) -> rstd::result::Result<ParaInfo, &'static str> {
+		let info = Paras::get(id).ok_or("invalid id")?;
+		if let Scheduling::Dynamic = info.scheduling {} else { Err("invalid parathread id")? }
+		Ok(info)
+	}
+}
+
+impl<T: Trait> ActiveParas for Module<T> {
+	fn active_paras() -> Vec<(ParaId, Option<CollatorId>, DispatchOrigins)> {
+		Active::get()
+	}
+}
+
+use srml_support::dispatch::IsSubType;
+
+/// Ensure that parathread selections happen prioritised by fees.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct LimitParathreadCommits<T: Trait>(rstd::marker::PhantomData<T>) where
+	T::Call: IsSubType<Module<T>, T>;
+
+#[cfg(feature = "std")]
+impl<T: Trait> rstd::fmt::Debug for LimitParathreadCommits<T> {
+	fn fmt(&self, f: &mut rstd::fmt::Formatter) -> rstd::fmt::Result {
+		write!(f, "LimitParathreadCommits<T>")
+	}
+}
+
+impl<T: Trait> SignedExtension for LimitParathreadCommits<T> {
+	type AccountId = T::AccountId;
+	type Call = T::Call;
+	type AdditionalSigned = ();
+	fn additional_signed(&self) -> rstd::result::Result<(), &'static str> { Ok(()) }
+	type Accumulator = Vec<ParaId>;
+
+	fn validate(
+		&self,
+		call: &Self::Call,
+		who: &Self::AccountId,
+		info: DispatchInfo,
+		len: usize,
+	) -> rstd::result::Result<ValidTransaction, DispatchError> {
+		if let Some(local_call) = call.is_aux_sub_type() {
+			if let Call::select_parathread(id, collator, hash) = local_call {
+				// ensure that the para ID is actually a parathread.
+				Self::ensure_thread_id(id)?;
+
+				// ensure that we haven't already had a full complement of selected parathreads.
+				let mut selected_threads = <SelectedThreads<T>>::get();
+				ensure!(selected_threads.len() < ThreadCount::get(), DispatchError::Overflow);
+
+				// ensure that this is not selecting a duplicate parathread ID
+				let pos = selected_threads
+					.binary_search_by(|&(ref other_id, _)| other_id == id)
+					.err()
+					.ok_or(DispatchError::BadState)?;
+
+				// ensure that this is a live bid (i.e. that the thread's chain head matches)
+				let actual = T::Hashing::hash(&<parachains::Module<T>>::parachain_head(id));
+				ensure!(actual == head_data_hash, DispatchError::Stale);
+
+				// updated the selected threads.
+				selected_threads.insert(pos, (id, collator));
```

Comment:
@tomusdrw don't know how well this will work with the current txqueue model... it assumes that we retain these changes in the candidate block's state *as we populate* our candidate block with txs from the tx queue.

Reply:
I don't fully understand how it relates to the pool, but if you call `validate` not only during `validate_transaction` but also during `apply_extrinsic` the changes will indeed be retained.
However when we import extrinsics to the pool they will runt `validate_transaction` on a pristine state of latest block and the changes are obviously not persisted.

Is that what you had in mind or did I miss your point completely?

Reply:
it should work ok like that, i think.

but generally, these transactions are a little more sophisticated than usual. there are several instances when they can become invalid:

1. the head-data-hash given in them is not (or no longer) the hash of the parathread's head-data. this can happen because:
  a. it was old when we received it (should never be imported into pool)
  b. it was an alternative fork (should never be imported into pool)
  c. the parathread has progressed since we received it (should be thrown out as soon as this happens).
2. two of them are for the same parathread with the same head-data-hash. in this case the one with the lower fee/priority should be discarded as the other will always supersede it.
3. some number (`registry::ThreadCount`) of parathread progressions have already occurred in this block. in this case, the transactions should remain in the pool as long as their head-data-hash has not become outdated.

as i understand it, the pool should import them as long as the HDH is current, it will correctly construct blocks with them properly prioritised (this just uses the fees priority) and it will avoid putting in more than it should into a block.

to make it proactively discard duplicates, we presumably need to ensure that their `provides` includes the same tag, which should fulfil the needs for (2).

but in order to get (1c) working, we'd need to essentially re-check the validity after each block is imported. does the queue do this already?